File: AOS/app/src/main/java/com/example/escapephone/core/theme/ThemeProgressStore.kt
package com.example.escapephone.core.theme

import android.content.Context
import com.example.escapephone.core.model.GameProgress
import com.example.escapephone.core.model.GameStage
import com.example.escapephone.core.persistence.SystemTimeProvider
import com.example.escapephone.core.persistence.TimeProvider
import kotlinx.serialization.SerializationException
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

interface ThemeProgressStore {
    fun loadThemeProgress(themeId: ThemeId): ThemeProgress
    fun saveThemeProgress(themeProgress: ThemeProgress)
    fun resetThemeProgress(themeId: ThemeId)
    fun resetAllThemeProgress()
    fun migrateLegacyProgressIfNeeded()
}

private val json = Json { ignoreUnknownKeys = true }

class PlatformThemeProgressStore(
    context: Context,
    private val timeProvider: TimeProvider = SystemTimeProvider
) : ThemeProgressStore {
    private val preferences = context.getSharedPreferences("escape_phone_themes", Context.MODE_PRIVATE)
    private val legacyPreferences = context.getSharedPreferences("escape_phone", Context.MODE_PRIVATE)

    override fun loadThemeProgress(themeId: ThemeId): ThemeProgress {
        val stored = preferences.getString(keyFor(themeId), null)
        val decoded = stored?.let { value ->
            try { json.decodeFromString<ThemeProgress>(value) } catch (_: SerializationException) { null } catch (_: IllegalArgumentException) { null }
        }
        return decoded ?: ThemeRegistry.defaultProgress(themeId)
    }

    override fun saveThemeProgress(themeProgress: ThemeProgress) {
        preferences.edit().putString(keyFor(themeProgress.themeId), json.encodeToString(themeProgress)).apply()
    }

    override fun resetThemeProgress(themeId: ThemeId) {
        preferences.edit().remove(keyFor(themeId)).apply()
    }

    override fun resetAllThemeProgress() {
        preferences.edit().clear().apply()
        preferences.edit().putBoolean(MIGRATION_FLAG_KEY, true).apply()
    }

    override fun migrateLegacyProgressIfNeeded() {
        if (preferences.getBoolean(MIGRATION_FLAG_KEY, false)) return
        preferences.edit().putBoolean(MIGRATION_FLAG_KEY, true).apply()

        val hasTheLastCommitProgress = preferences.contains(keyFor(ThemeId.theLastCommit))
        if (hasTheLastCommitProgress) return

        val legacyRaw = legacyPreferences.getString("escape_phone_game_progress_v1", null) ?: return
        val legacy = try { json.decodeFromString<GameProgress>(legacyRaw) } catch (_: SerializationException) { null } catch (_: IllegalArgumentException) { null } ?: return

        val migrated = migrateLegacyGameProgress(legacy, timeProvider)
        saveThemeProgress(migrated)
    }

    private fun keyFor(themeId: ThemeId) = "theme_progress_${themeId.storageValue}"

    companion object { const val MIGRATION_FLAG_KEY = "legacy_migration_done" }
}

fun migrateLegacyGameProgress(legacy: GameProgress, timeProvider: TimeProvider = SystemTimeProvider): ThemeProgress {
    val status = when {
        legacy.currentStage == GameStage.gameCompleted -> ThemeStatus.completed
        legacy.currentStage != GameStage.notStarted -> ThemeStatus.inProgress
        else -> ThemeStatus.notStarted
    }
    return ThemeProgress(
        themeId = ThemeId.theLastCommit,
        themeStatus = status,
        currentStageId = legacy.currentStage.name,
        startedAt = legacy.startedAt,
        lastSavedAt = legacy.lastSavedAt ?: timeProvider.now(),
        completedAt = legacy.completedAt,
        hintCount = legacy.hintCount,
        collectedEvidenceIds = legacy.collectedEvidenceIds,
        endingId = legacy.endingType?.name,
        themeSpecificState = ThemeSpecificState.TheLastCommitState()
    )
}

class InMemoryThemeProgressStore(
    private val progressByTheme: MutableMap<ThemeId, ThemeProgress> = mutableMapOf(),
    private var legacyGameProgress: GameProgress? = null,
    private val timeProvider: TimeProvider = SystemTimeProvider
) : ThemeProgressStore {
    private var migrationDone = false

    override fun loadThemeProgress(themeId: ThemeId): ThemeProgress = progressByTheme[themeId] ?: ThemeRegistry.defaultProgress(themeId)
    override fun saveThemeProgress(themeProgress: ThemeProgress) { progressByTheme[themeProgress.themeId] = themeProgress }
    override fun resetThemeProgress(themeId: ThemeId) { progressByTheme.remove(themeId) }
    override fun resetAllThemeProgress() { progressByTheme.clear(); migrationDone = true }
    override fun migrateLegacyProgressIfNeeded() {
        if (migrationDone) return
        migrationDone = true
        if (progressByTheme.containsKey(ThemeId.theLastCommit)) return
        val legacy = legacyGameProgress ?: return
        progressByTheme[ThemeId.theLastCommit] = migrateLegacyGameProgress(legacy, timeProvider)
    }

    fun setLegacyGameProgress(gameProgress: GameProgress?) { legacyGameProgress = gameProgress }
}

File: ios/EscapePhone/Core/Theme/ThemeProgressStore.swift
import Foundation

protocol ThemeProgressStore {
    func loadThemeProgress(themeId: ThemeId) -> ThemeProgress
    func saveThemeProgress(_ themeProgress: ThemeProgress)
    func resetThemeProgress(themeId: ThemeId)
    func resetAllThemeProgress()
    func migrateLegacyProgressIfNeeded()
}

func migrateLegacyGameProgress(_ legacy: GameProgress, now: Date) -> ThemeProgress {
    let status: ThemeStatus
    switch legacy.currentStage {
    case .gameCompleted: status = .completed
    case .notStarted: status = .notStarted
    default: status = .inProgress
    }
    return ThemeProgress(
        themeId: .theLastCommit,
        themeStatus: status,
        currentStageId: legacy.currentStage.rawValue,
        startedAt: legacy.startedAt,
        lastSavedAt: legacy.lastSavedAt ?? now,
        completedAt: legacy.completedAt,
        hintCount: legacy.hintCount,
        collectedEvidenceIds: legacy.collectedEvidenceIds,
        endingId: legacy.endingType?.rawValue,
        themeSpecificState: .theLastCommit(.init())
    )
}

final class PlatformThemeProgressStore: ThemeProgressStore {
    private let defaults: UserDefaults
    private let legacyKey: String
    private let dateProvider: DateProviding
    private let migrationFlagKey = "legacy_migration_done"

    init(defaults: UserDefaults = .standard, legacyKey: String = "escape_phone_game_progress_v1", dateProvider: DateProviding = SystemDateProvider()) {
        self.defaults = defaults; self.legacyKey = legacyKey; self.dateProvider = dateProvider
    }

    func loadThemeProgress(themeId: ThemeId) -> ThemeProgress {
        guard let data = defaults.data(forKey: key(for: themeId)), let decoded = try? JSONDecoder.themeDecoder().decode(ThemeProgress.self, from: data) else {
            return ThemeRegistry.defaultProgress(for: themeId)
        }
        return decoded
    }

    func saveThemeProgress(_ themeProgress: ThemeProgress) {
        guard let data = try? JSONEncoder.themeEncoder().encode(themeProgress) else { return }
        defaults.set(data, forKey: key(for: themeProgress.themeId))
    }

    func resetThemeProgress(themeId: ThemeId) { defaults.removeObject(forKey: key(for: themeId)) }

    func resetAllThemeProgress() {
        for theme in ThemeId.allCases { defaults.removeObject(forKey: key(for: theme)) }
        defaults.set(true, forKey: migrationFlagKey)
    }

    func migrateLegacyProgressIfNeeded() {
        guard !defaults.bool(forKey: migrationFlagKey) else { return }
        defaults.set(true, forKey: migrationFlagKey)
        guard defaults.data(forKey: key(for: .theLastCommit)) == nil else { return }
        guard let legacyData = defaults.data(forKey: legacyKey), let legacy = try? JSONDecoder().decode(GameProgress.self, from: legacyData) else { return }
        saveThemeProgress(migrateLegacyGameProgress(legacy, now: dateProvider.now))
    }

    private func key(for themeId: ThemeId) -> String { "theme_progress_\(themeId.rawValue)" }
}

final class InMemoryThemeProgressStore: ThemeProgressStore {
    private var progressByTheme: [ThemeId: ThemeProgress]
    private var legacyGameProgress: GameProgress?
    private let dateProvider: DateProviding
    private var migrationDone = false

    init(progressByTheme: [ThemeId: ThemeProgress] = [:], legacyGameProgress: GameProgress? = nil, dateProvider: DateProviding = SystemDateProvider()) {
        self.progressByTheme = progressByTheme; self.legacyGameProgress = legacyGameProgress; self.dateProvider = dateProvider
    }

    func loadThemeProgress(themeId: ThemeId) -> ThemeProgress { progressByTheme[themeId] ?? ThemeRegistry.defaultProgress(for: themeId) }
    func saveThemeProgress(_ themeProgress: ThemeProgress) { progressByTheme[themeProgress.themeId] = themeProgress }
    func resetThemeProgress(themeId: ThemeId) { progressByTheme.removeValue(forKey: themeId) }
    func resetAllThemeProgress() { progressByTheme.removeAll(); migrationDone = true }
    func migrateLegacyProgressIfNeeded() {
        guard !migrationDone else { return }
        migrationDone = true
        guard progressByTheme[.theLastCommit] == nil else { return }
        guard let legacy = legacyGameProgress else { return }
        progressByTheme[.theLastCommit] = migrateLegacyGameProgress(legacy, now: dateProvider.now)
    }

    func setLegacyGameProgress(_ gameProgress: GameProgress?) { legacyGameProgress = gameProgress }
}

extension JSONEncoder {
    static func themeEncoder() -> JSONEncoder { let e = JSONEncoder(); e.dateEncodingStrategy = .millisecondsSince1970; return e }
}
extension JSONDecoder {
    static func themeDecoder() -> JSONDecoder { let d = JSONDecoder(); d.dateDecodingStrategy = .millisecondsSince1970; return d }
}
